### PR TITLE
Subgraph isomorphism algorithm addition

### PR DESCRIPTION
**Pull request by orazve**
Added subgraph isomorphism algorithm

### CODE REVIEW

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/si.hpp [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+#include "oneapi/dal/algo/subgraph_isomorphism/backend/cpu/sorter.hpp"
+#include "oneapi/dal/algo/subgraph_isomorphism/backend/cpu/matching.hpp"
+#include "oneapi/dal/algo/subgraph_isomorphism/common.hpp"
+#include "oneapi/dal/algo/subgraph_isomorphism/graph_matching_types.hpp"
+#include "oneapi/dal/backend/dispatcher.hpp"
+
+namespace oneapi::dal::preview::subgraph_isomorphism::backend {
+using namespace oneapi::dal::preview::subgraph_isomorphism::backend;
+
+template <typename Cpu = oneapi::dal::backend::cpu_dispatch_sse2>
```
Comment (orrrrtem) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/si.hpp:30]: Why default is set? Why ssse2?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/bit_vector.cpp [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/algo/subgraph_isomorphism/backend/cpu/bit_vector.hpp"
+
+namespace oneapi::dal::preview::subgraph_isomorphism::backend {
+
+void or_equal(std::uint8_t* ONEAPI_RESTRICT vec,
```
Comment (orrrrtem) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/bit_vector.cpp:21]: Template CPU?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/bit_vector.cpp [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+              const std::uint8_t* ONEAPI_RESTRICT pa,
+              std::int64_t size) {
+    ONEDAL_IVDEP
+    for (std::int64_t i = 0; i < size; i++) {
+        vec[i] |= pa[i];
+    }
+}
+void and_equal(std::uint8_t* ONEAPI_RESTRICT vec,
```
Comment (orrrrtem) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/bit_vector.cpp:29]: The same question for each part of backend without CPU. Illegal instruction and optimization dependency  may appear (from machine where library is built at).

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/bit_vector.cpp [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+void bit_vector::set(const std::uint8_t byte_val) {
+    const std::int64_t nn = n;
+    ONEDAL_VECTOR_ALWAYS
+    for (std::int64_t i = 0; i < nn; i++) {
+        vector[i] = byte_val;
+    }
+}
+
+graph_status bit_vector::set_bit(std::uint8_t* result_vector, const std::int64_t vertex) {
+    if (result_vector == nullptr) {
```
Comment (orrrrtem) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/bit_vector.cpp:128]: Do we need such check inside set_bit operation? Can we check it outside?

**Review by orrrrtem — COMMENTED** [review]
Initial part of comments

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp [hunk 5]
```diff
@@ -0,0 +1,648 @@
...
+
+    return p_edges_attribute[current_vertex][attribute_index];
+}
+
+template <typename Cpu>
+graph_input_data<Cpu>::graph_input_data(inner_alloc allocator) : allocator_(allocator) {
+    vertex_count = 0;
+    degree = nullptr;
+    attr = nullptr;
+    edges_attribute = nullptr;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp:513]: Initialization

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp [hunk 6]
```diff
@@ -0,0 +1,648 @@
...
+template <typename Cpu>
+graph_input_data<Cpu>::graph_input_data(const std::int64_t vertex_size, inner_alloc allocator)
+        : allocator_(allocator) {
+    vertex_count = vertex_size;
+
+    degree = allocator_.allocate<std::int64_t>(vertex_count);
+    attr = allocator_.allocate<std::int64_t>(vertex_count);
+
+    edges_attribute = allocator_.allocate<std::int64_t*>(vertex_count);
+    if (edges_attribute != nullptr) {
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp:525]: Very unclear

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp [hunk 7]
```diff
@@ -0,0 +1,648 @@
...
+        for (int64_t i = 0; i < vertex_count; i++) {
+            edges_attribute[i] = nullptr;
+            degree[i] = 0;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp:528]: SDL

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp [hunk 8]
```diff
@@ -0,0 +1,648 @@
...
+            attr[i] = 1;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp:529]: SDL

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp [hunk 9]
```diff
@@ -0,0 +1,648 @@
...
+            edges_attribute[i] = nullptr;
+        }
+    }
+    allocator_.deallocate<std::int64_t*>(edges_attribute, vertex_count);
+}
+
+template <typename Cpu>
+graph_input_list_data<Cpu>::graph_input_list_data(inner_alloc allocator)
+        : graph_input_data<Cpu>(allocator) {
+    data = nullptr;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp:551]: Initialization

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp [hunk 10]
```diff
@@ -0,0 +1,648 @@
...
+}
+
+template <typename Cpu>
+graph_input_list_data<Cpu>::graph_input_list_data(graph_input_data<Cpu>* input_data,
+                                                  inner_alloc allocator)
+        : graph_input_data<Cpu>(allocator) {
+    vertex_count = input_data->vertex_count;
+    degree = input_data->degree;
+    attr = input_data->attr;
+    edges_attribute = input_data->edges_attribute;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp:571]: Inititialization

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp [hunk 11]
```diff
@@ -0,0 +1,648 @@
...
+            data[i] = nullptr;
+        }
+    }
+    allocator_.template deallocate<std::int64_t*>(data, vertex_count);
+}
+
+template <typename Cpu>
+graph_input_bit_data<Cpu>::graph_input_bit_data(inner_alloc allocator)
+        : graph_input_data<Cpu>(allocator) {
+    data = nullptr;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp:598]: init

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp [hunk 12]
```diff
@@ -0,0 +1,648 @@
...
+}
+
+template <typename Cpu>
+graph_input_bit_data<Cpu>::graph_input_bit_data(const std::int64_t vertex_size,
+                                                inner_alloc allocator)
+        : graph_input_data<Cpu>(vertex_size, allocator) {
+    std::int64_t bit_array_size = bit_vector<Cpu>::bit_vector_size(vertex_count);
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp:605]: const

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp [hunk 13]
```diff
@@ -0,0 +1,648 @@
...
+}
+
+template <typename Cpu>
+graph_input_bit_data<Cpu>::graph_input_bit_data(graph_input_data<Cpu>* input_data,
+                                                inner_alloc allocator)
+        : graph_input_data<Cpu>(allocator) {
+    vertex_count = input_data->vertex_count;
+    degree = input_data->degree;
+    attr = input_data->attr;
+    edges_attribute = input_data->edges_attribute;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp:621]: init

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp [hunk 14]
```diff
@@ -0,0 +1,648 @@
...
+
+    input_data->vertex_count = 0;
+    input_data->degree = nullptr;
+    input_data->attr = nullptr;
+    input_data->edges_attribute = nullptr;
+
+    std::int64_t bit_array_size = bit_vector<Cpu>::bit_vector_size(vertex_count);
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp:628]: const

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp [hunk 15]
```diff
@@ -0,0 +1,648 @@
...
+
+    data = allocator_.template allocate<std::uint8_t*>(vertex_count);
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph.hpp:630]: init

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph_status.hpp [hunk 16]
```diff
@@ -0,0 +1,32 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph_status.hpp:2]: 2021

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph_status.hpp [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace oneapi::dal::preview::subgraph_isomorphism::backend {
+
+enum graph_status {
+    ok = 0, /*!< No error found*/
+    bad_arguments = -5, /*!< Bad argument(s) passed*/
+    bad_allocation = -11, /*!< Memory allocation error*/
+};
+
+enum edge_direction {
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/graph_status.hpp:27]: How it related to graph status?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/inner_alloc.hpp [hunk 18]
```diff
@@ -0,0 +1,81 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/inner_alloc.hpp:2]: 2021

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/inner_alloc.hpp [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+namespace oneapi::dal::preview::subgraph_isomorphism::detail {
+struct byte_alloc_iface;
+}
+
+#if defined(__INTEL_COMPILER)
+#define ONEAPI_RESTRICT
+//restrict
+#else
+#define ONEAPI_RESTRICT
+#endif
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/inner_alloc.hpp:32]: Looks wrong

**Review by averbukh — COMMENTED** [review]
Reviewed up to matching.hpp (too many files). I plan to continue.

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/debug.cpp [hunk 20]
```diff
@@ -0,0 +1,59 @@
+/*******************************************************************************
```
[the remark this answers is not in the document]
Comment (orazve) [reply cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/debug.cpp:1]: yes

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/gcc_adapt.hpp [hunk 21]
```diff
@@ -0,0 +1,75 @@
+/*******************************************************************************
```
[the remark this answers is not in the document]
Comment (orazve) [reply cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/gcc_adapt.hpp:1]: yes, the file will be renamed

File: examples/oneapi/cpp/BUILD [hunk 22]
```diff
@@ -8,6 +8,9 @@ dal_module(
     name = "example_util",
     hdrs = glob(["source/example_util/*.hpp"]),
     includes = [ "source" ],
+    dal_deps = [
+        "@onedal//cpp/oneapi/dal/algo:jaccard",
```
Comment (bysheva) [inline examples/oneapi/cpp/BUILD:12]: jaccard?

File: examples/oneapi/cpp/source/example_util/output_helpers.hpp [hunk 23]
```diff
@@ -55,3 +55,42 @@ std::ostream &operator<<(std::ostream &stream, const oneapi::dal::table &table)
...
+void print_table_int(const oneapi::dal::table &table) {
+    if (!table.has_data())
+        return;
+    auto arr = oneapi::dal::row_accessor<const int>(table).pull();
+    const auto x = arr.get_data();
+
+    for (std::int64_t i = 0; i < table.get_row_count(); i++) {
+        for (std::int64_t j = 0; j < table.get_column_count(); j++) {
+            std::cout << x[i * table.get_column_count() + j];
+            // if (j < (table.get_column_count() - 1))
```
Comment (bysheva) [inline examples/oneapi/cpp/source/example_util/output_helpers.hpp:68]: remove

File: examples/oneapi/cpp/source/subgraph_isomorphism/subgraph_isomorphism_batch.cpp [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+            .set_semantic_match(false)
+            .set_max_match_count(100);
+
+    const auto result =
+        dal::preview::graph_matching(subgraph_isomorphism_desc, target_graph, pattern_graph);
+
+    // extract the result
+    std::cout << "Number of matchings: " << result.get_match_count() << std::endl;
+    std::cout << "Matchings:\n" << result.get_vertex_match() << std::endl;
+    // print_table_int_sorted(result.get_vertex_match());
```
Comment (bysheva) [inline examples/oneapi/cpp/source/subgraph_isomorphism/subgraph_isomorphism_batch.cpp:59]: Remove

**Review by bysheva — CHANGES_REQUESTED** [review]

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/debug.cpp [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/algo/subgraph_isomorphism/backend/cpu/debug.hpp"
+
+namespace oneapi::dal::preview::subgraph_isomorphism::backend {
```
[the remark this answers is not in the document]
Comment (orazve) [reply cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/debug.cpp]: removed

File: cpp/oneapi/dal/algo/BUILD [hunk 26]
```diff
@@ -30,3 +32,11 @@ dal_collect_test_suites(
...
     modules = ALGOS,
 )
+
+dal_module(
+    name = "graph_headers",
+    hdrs = [
+        "subgraph_isomorphism.hpp",
+    ],
+    includes = ["."],
+)
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/BUILD:42]: Why this module is needed?

File: examples/oneapi/cpp/BUILD (hunk 22 above)
Comment (rlnx) [reply examples/oneapi/cpp/BUILD:12]: Yes, it looks strange, to build, e.g., PCA example, I'll need to build jaccard

File: cpp/oneapi/dal/algo/BUILD (hunk 26 above)
Comment (orazve) [reply cpp/oneapi/dal/algo/BUILD]: fixed, removed

**Review by bysheva — APPROVED** [review]

**Pull request merged**